Extend/Controller/OrderController.php

```diff
@@ -232,22 +232,37 @@ private function _handleTransaction($oBasket, $oOrder)
      * @param Order           $oOrder
      *
      * @return void
+     *
+     * @since 1.0.0
      */
     private function _handleResponse($oResponse, $oLogger, $oOrder)
     {
         if ($oResponse instanceof FailureResponse) {
             $oLogger->error('Error processing transaction:');
 
+            $aErrorDescriptions = [];
+
             foreach ($oResponse->getStatusCollection() as $oStatus) {
                 /**
                  * @var Status $oStatus
                  */
                 $sSeverity = ucfirst($oStatus->getSeverity());
                 $sCode = $oStatus->getCode();
                 $sDescription = $oStatus->getDescription();
+
+                // add error message and code to array for display on frontend
+                $aErrorDescriptions[] = $sDescription . ' (Error code: ' . $sCode . ')';
+
                 $oLogger->error("\t$sSeverity with code $sCode and message '$sDescription' occurred.");
             }
-            return;
+
+            // set the custom payment error code and text and redirect back to the payment step of the checkout process
+            Registry::getSession()->setVariable('payerror', '-102');
+            Registry::getSession()->setVariable('payerrortext', join('<br/>', $aErrorDescriptions));
+
+            $sRedirectUrl = Registry::getConfig()->getShopHomeUrl() . 'cl=payment';
```

Comment:
just to be sure, let's add the session id too.

Reply:
`Registry::getConfig()->getShopHomeUrl()` returns the shop base URL + index.php + the sid automatically.

Reply:
If that's the case, don't you need to use an ampersand (`&`) before `cl=payment`, to join with the other params?